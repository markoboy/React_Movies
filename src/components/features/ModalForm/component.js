import Modal from '@components/common/Modal';
import PropTypes from 'prop-types';
import React from 'react';

export default function ModalFormComponent({
  onSubmit,
  onReset,
  title,
  onCloseTrigger,
  body,
  footer,
}) {
  return (
    <form onSubmit={onSubmit} onReset={onReset}>
      <Modal
        title={title}
        onCloseTrigger={onCloseTrigger}
        body={body}
        footer={footer}
      />
    </form>
  );
}

ModalFormComponent.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onReset: PropTypes.func.isRequired,
  onCloseTrigger: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  body: PropTypes.element.isRequired,
  footer: PropTypes.element.isRequired,
};
