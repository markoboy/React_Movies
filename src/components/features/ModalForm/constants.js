export const MODAL_ADD_TITLE = 'Add Movie';
export const MODAL_EDIT_TITLE = 'Edit Movie';
export const MODAL_DELETE_TITLE = 'Delete Movie';

export const MODAL_FORM_ADD_ACTION = 'Add';
export const MODAL_FORM_EDIT_ACTION = 'Edit';
export const MODAL_FORM_DELETE_ACTION = 'Delete';

export const MODAL_FORM_DELETE_PARAGRAPH = 'Are you sure you want to delete this movie?';

export const GENRE_SELECT_OPTIONS = [
  {
    label: 'Documentary',
    value: 'documentary',
  },
  {
    label: 'Comedy',
    value: 'comedy',
  },
  {
    label: 'Horror',
    value: 'horror',
  },
  {
    label: 'Crime',
    value: 'crime',
  },
];
