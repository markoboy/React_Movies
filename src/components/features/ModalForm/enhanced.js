import { resetModalStateCreator } from '@store/action-creators/modalActionCreators';
import {
  resetModalFormStateCreator,
  setFormInputsCreator,
} from '@store/action-creators/modalFormActionCreators';
import {
  modalFormActionSelector,
  modalFormInputsSelector,
  modalFormSelectedMovieSelector,
} from '@store/selectors/modalFormSelectors';
import { modalTitleSelector } from '@store/selectors/modalSelectors';
import { addMovie, deleteMovie, updateMovie } from '@store/thunks/moviesThunk';
import { connect } from 'react-redux';
import ModalFormContainer from './container';

const mapStateToProps = (state) => ({
  title: modalTitleSelector(state),
  formAction: modalFormActionSelector(state),
  formInputs: modalFormInputsSelector(state),
  selectedMovie: modalFormSelectedMovieSelector(state),
});

const mapDispatchToProps = {
  resetModalState: resetModalStateCreator,
  resetModalFormState: resetModalFormStateCreator,
  setFormInputs: setFormInputsCreator,
  deleteMovie,
  updateMovie,
  addMovie,
};

export default connect(mapStateToProps, mapDispatchToProps)(ModalFormContainer);
