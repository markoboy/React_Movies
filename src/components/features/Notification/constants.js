export const SUCCESS_TITLE = 'Congratulations!';
export const ERROR_TITLE = 'Oops an Error occurred!';
