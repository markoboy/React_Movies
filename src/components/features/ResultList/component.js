import PropTypes from 'prop-types';
import React from 'react';
import ResultCount from './ResultCount';

function ResultListComponent({ totalAmount, children }) {
  return (
    <>
      <ResultCount count={totalAmount} />
      <section className="section section--padding-bottom-only">
        <ul className="result-list">{children}</ul>
      </section>
    </>
  );
}

ResultListComponent.propTypes = {
  totalAmount: PropTypes.number.isRequired,
};

export default ResultListComponent;
