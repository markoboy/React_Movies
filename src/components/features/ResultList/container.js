import {
  MODAL_FORM_DELETE_ACTION,
  MODAL_FORM_EDIT_ACTION,
} from '@components/features/ModalForm/constants';
import ResultItemPopOut from '@components/features/ResultList/ResultItemPopOut';
import ResultListItem from '@components/features/ResultList/ResultListItem';
import PropTypes from 'prop-types';
import React, { useCallback } from 'react';
import ResultListComponent from './component';

const tileActions = [MODAL_FORM_EDIT_ACTION, MODAL_FORM_DELETE_ACTION];

export default function ResultListContainer({
  movies,
  totalAmount,
  setFormSelectedMovie,
  setFormAction,
  setIsOpenedModal,
  selectMovie,
}) {
  const handleActionClick = useCallback((action, movie) => {
    setFormSelectedMovie(movie);
    setFormAction(action);
    setIsOpenedModal(true);
  }, []);

  const handleResultItemClick = useCallback((movieId) => {
    selectMovie(movieId);
    window.document.body.scrollTo({ top: 0, behavior: 'smooth' });
  }, []);

  const getResultListItems = useCallback(
    (movie) => (
      <ResultListItem
        key={movie.id}
        action={movie.id}
        onClick={handleResultItemClick}
        {...movie}
      >
        <ResultItemPopOut
          movie={movie}
          actions={tileActions}
          onClick={handleActionClick}
        />
      </ResultListItem>
    ),
    [movies]
  );

  return (
    <ResultListComponent totalAmount={totalAmount}>
      {movies.map(getResultListItems)}
    </ResultListComponent>
  );
}

ResultListContainer.propTypes = {
  movies: PropTypes.arrayOf(PropTypes.object).isRequired,
  totalAmount: PropTypes.number.isRequired,
  setFormSelectedMovie: PropTypes.func.isRequired,
  setFormAction: PropTypes.func.isRequired,
  setIsOpenedModal: PropTypes.func.isRequired,
  selectMovie: PropTypes.func.isRequired,
};
