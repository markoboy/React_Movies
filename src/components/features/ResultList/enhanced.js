import withWrapper from '@components/hocs/WithWrapper';
import { setIsOpenedModalCreator } from '@store/action-creators/modalActionCreators';
import {
  setFormActionCreator,
  setFormSelectedMovieCreator,
} from '@store/action-creators/modalFormActionCreators';
import {
  moviesSelector,
  moviesTotalAmountSelector,
} from '@store/selectors/moviesSelectors';
import { memo } from 'react';
import { connect } from 'react-redux';
import ResultListContainer from './container';

const mapStateToProps = (state) => ({
  movies: moviesSelector(state),
  totalAmount: moviesTotalAmountSelector(state),
});

const mapDispatchToProps = {
  setFormSelectedMovie: setFormSelectedMovieCreator,
  setFormAction: setFormActionCreator,
  setIsOpenedModal: setIsOpenedModalCreator,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(memo(withWrapper(ResultListContainer)));
