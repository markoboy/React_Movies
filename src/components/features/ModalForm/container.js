import {
  MODAL_FORM_ADD_ACTION,
  MODAL_FORM_DELETE_ACTION,
  MODAL_FORM_EDIT_ACTION,
} from '@components/features/ModalForm/constants';
import { serializeMovieData } from '@components/features/ModalForm/utils';
import { MovieDetailType } from '@constants/MovieTypes';
import LoggerService from '@services/LoggerService';
import PropTypes from 'prop-types';
import React, { useCallback } from 'react';
import ModalFormComponent from './component';
import ModalFormBody from './ModalFormBody';

const modalFormLogger = new LoggerService('ModalFormContainer');

export default function ModalFormContainer({
  title,
  formAction,
  selectedMovie,
  resetModalState,
  resetModalFormState,
  deleteMovie,
  updateMovie,
  addMovie,
}) {
  const handleCloseTrigger = useCallback(() => {
    resetModalState();
    resetModalFormState();
  }, []);

  const handleOnSubmit = useCallback(
    /**
     * @param {import('formik').FormikValues} movieFormValues
     * @param {import('formik').FormikHelpers} formikHelpers
     */
    (movieFormValues, formikHelpers) => {
      switch (formAction) {
        case MODAL_FORM_DELETE_ACTION:
          deleteMovie(selectedMovie.id);
          break;

        case MODAL_FORM_EDIT_ACTION:
          updateMovie(serializeMovieData(movieFormValues));
          break;

        case MODAL_FORM_ADD_ACTION:
          addMovie(serializeMovieData(movieFormValues));
          break;

        default:
          modalFormLogger.warn(
            `FormAction: ${formAction} doesnt exist on modal form submit!`
          );
          break;
      }

      if (formikHelpers) {
        formikHelpers.setSubmitting(false);
      }

      handleCloseTrigger();
    },
    [formAction, selectedMovie]
  );

  return (
    <ModalFormComponent
      onCloseTrigger={handleCloseTrigger}
      title={title}
      body={<ModalFormBody onSubmit={handleOnSubmit} />}
    />
  );
}

ModalFormContainer.defaultProps = {
  formAction: null,
  selectedMovie: null,
};

ModalFormContainer.propTypes = {
  title: PropTypes.string.isRequired,

  formAction: PropTypes.string,
  selectedMovie: PropTypes.shape(MovieDetailType),

  resetModalState: PropTypes.func.isRequired,
  resetModalFormState: PropTypes.func.isRequired,

  deleteMovie: PropTypes.func.isRequired,
  updateMovie: PropTypes.func.isRequired,
  addMovie: PropTypes.func.isRequired,
};
