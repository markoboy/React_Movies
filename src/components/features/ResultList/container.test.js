import React from 'react';
import userEvent from '@testing-library/user-event';

import ResultList from './container';
import { renderWithRouter } from '@utils/testUtils';
import { MODAL_FORM_EDIT_ACTION } from '../ModalForm/constants';

const movies = [
  {
    id: 1,
    title: 'A film title',
    poster_path: 'https://film.com/image.jpg',
    overview: 'A great film',
    runtime: 145,
    genres: ['Documentary'],
    release_date: new Date(),
  },
];

const totalAmount = movies.length;

describe('ResultList Component', () => {
  let resultListRender;

  let setFormSelectedMovie;
  let setFormAction;
  let setIsOpenedModal;

  beforeEach(() => {
    setFormSelectedMovie = jest.fn();
    setFormAction = jest.fn();
    setIsOpenedModal = jest.fn();

    resultListRender = renderWithRouter(
      <ResultList
        movies={movies}
        totalAmount={totalAmount}
        setFormSelectedMovie={setFormSelectedMovie}
        setFormAction={setFormAction}
        setIsOpenedModal={setIsOpenedModal}
      />
    );
  });

  it('renders the correct amount of movies', () => {
    const { getAllByTestId } = resultListRender;

    const items = getAllByTestId(/result-list-item/i);

    expect(items.length).toBe(movies.length);
  });

  it('triggers setFormSelectedMovie, setFormAction and setIsOpenedModal when an action is clicked', () => {
    const { getAllByText } = resultListRender;

    const editActions = getAllByText(MODAL_FORM_EDIT_ACTION);

    expect(editActions.length).toBe(movies.length);

    userEvent.click(editActions[0]);

    expect(setFormSelectedMovie).toHaveBeenCalledTimes(1);
    expect(setFormAction).toHaveBeenCalledTimes(1);
    expect(setIsOpenedModal).toHaveBeenCalledTimes(1);

    expect(setFormSelectedMovie).toHaveBeenCalledWith(movies[0]);
    expect(setFormAction).toHaveBeenCalledWith(MODAL_FORM_EDIT_ACTION);
    expect(setIsOpenedModal).toHaveBeenCalledWith(true);
  });
});
