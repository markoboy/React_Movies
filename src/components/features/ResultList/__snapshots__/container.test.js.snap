// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultList Component renders the correct amount of movies 1`] = `
<div>
  <p>
    <span
      class="font--bold"
    >
      1
    </span>
     
    movies found
  </p>
  <section
    class="section section--padding-bottom-only"
  >
    <ul
      class="result-list"
    >
      <li
        class="result-list__item"
        data-testid="result-list-item"
      >
        <a
          class="result-item-container"
          href="/film/1"
        >
          <img
            alt="A film title"
            src="https://film.com/image.jpg"
          />
          <div
            class="result-item__body flex flex--wrap"
          >
            <h4
              class="result-item__title"
            >
              A film title
            </h4>
            <p
              class="result-item__release-date"
            >
              2021
            </p>
            <p
              class="result-item__genre"
            >
              Documentary
            </p>
          </div>
        </a>
        <div
          class="result-item__pop-out"
        >
          <button
            class="btn pop-out__btn"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-ellipsis-v fa-w-6 fa-lg "
              data-icon="ellipsis-v"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 192 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                fill="currentColor"
              />
            </svg>
          </button>
          <div
            class="pop-out__content"
          >
            <button
              class="btn btn--full-width btn--secondary pop-out-content__item"
              tabindex="0"
              type="button"
            >
              Edit
            </button>
            <button
              class="btn btn--full-width btn--secondary pop-out-content__item"
              tabindex="0"
              type="button"
            >
              Delete
            </button>
          </div>
        </div>
      </li>
    </ul>
  </section>
</div>
`;
